transports/quic/src/transport.rs

```diff
@@ -0,0 +1,641 @@
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Implementation of the [`Transport`] trait for QUIC.
+//!
+//! Combines all the objects in the other modules to implement the trait.
+
+use crate::connection::Connection;
+use crate::endpoint::ToEndpoint;
+use crate::{endpoint::EndpointChannel, muxer::QuicMuxer, upgrade::Upgrade};
+use crate::{Config, ConnectionError};
+
+#[cfg(feature = "async-std")]
+mod async_std;
+#[cfg(feature = "tokio")]
+mod tokio;
+#[cfg(feature = "async-std")]
+pub use async_std::{AsyncStd, AsyncStdTransport};
+use futures::future::BoxFuture;
+#[cfg(feature = "tokio")]
+pub use tokio::{Tokio, TokioTransport};
+
+use futures::channel::{mpsc, oneshot};
+use futures::ready;
+use futures::stream::StreamExt;
+use futures::{prelude::*, stream::SelectAll};
+
+use if_watch::{IfEvent, IfWatcher};
+
+use libp2p_core::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerId, TransportError as CoreTransportError, TransportEvent},
+    PeerId, Transport,
+};
+use std::collections::VecDeque;
+use std::io;
+use std::marker::PhantomData;
+use std::net::{IpAddr, UdpSocket};
+use std::task::Waker;
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[derive(Debug)]
+pub struct QuicTransport<P> {
+    config: Config,
+    listeners: SelectAll<Listener>,
+    /// Dialer for Ipv4 addresses if no matching listener exists.
+    ipv4_dialer: Option<Dialer>,
+    /// Dialer for Ipv6 addresses if no matching listener exists.
+    ipv6_dialer: Option<Dialer>,
+
+    _marker: PhantomData<P>,
+}
+
+impl<P> QuicTransport<P> {
+    pub fn new(config: Config) -> Self {
+        Self {
+            listeners: SelectAll::new(),
+            config,
+            ipv4_dialer: None,
+            ipv6_dialer: None,
+            _marker: Default::default(),
+        }
+    }
+}
+
+/// Error that can happen on the transport.
+#[derive(Debug, thiserror::Error)]
+pub enum TransportError {
+    /// Error while trying to reach a remote.
+    #[error("{0}")]
+    Reach(quinn_proto::ConnectError),
```

Comment:
Errors should not forward their `Display` impl to their inner error. See https://github.com/rust-lang/project-error-handling/issues/44.

I'd be better to use `#[source]` or `#[from]` and put `#[error(transparent)]` on this variant if there no error text that you want to add to the error chain here.

Reply:
Suggestion included in 6f284cc.